### PR TITLE
Improve image pasting from MS Word

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #12849

When pasting an image from MS Word or LibreOffice Writer, the images were being inserted with one of two formats for their source attributes (Seemed random which it was):
1. A `file://` URI pointing to the image in a temporary directory
2. A `blob:` URI that seemed to point to a broken image

Local file access is not allowed in at least Chrome and Firefox for security reasons. However, the raw clipboard data has 3 different data sources (plain text, HTML, and RTF). TinyMCE only uses HTML or plain text but the RTF source has the binary image contents in hexadecimal. We can extract that data and replace the broken image sources.

I don't know if this fix would cause issues with other copy-paste cases that worked previously but I've not found any issues yet.
There is still an error printed in the browser console as it tries to initially access the local file. I haven't found a way to replace the `src` value earlier to prevent that error.

### CODE REVIEW

File: js/fileupload.js [hunk 1]
```diff
@@ -353,13 +353,52 @@ var insertImageInTinyMCE = function(editor, image) {
...
+            }
             // Trigger upload process for each pasted image
             var fragment = $('<div></div>');
             fragment.append(event.content);
             fragment.find('img').each(function() {
                 const image = $(this);
-                const src = image.attr('src');
+                let src = image.attr('src');
+                const file_pattern = '^file://';
+                const local_file_blob_pattern = `^blob:${CFG_GLPI.url_base}/[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}`;
```
Comment (cedric-anne) [inline js/fileupload.js:394]: This kind of URL is produced by `URL.createObjectURL(file)` (line 95).

I think the only way to reproduce presence of this kind of URLs is to copy the data I just pasted and paste it again in the editor. I propose to not handle this case in this PR, as it a specific case that will not be a problem anymore in another fix I am working on in order to handle #12961.

File: js/fileupload.js [hunk 2]
```diff
@@ -353,13 +353,52 @@ var insertImageInTinyMCE = function(editor, image) {
...
+        const rtf_img_types = {
+            'pngblip': 'image/png',
+            'jpegblip': 'image/jpeg',
+        };
+        editor.on('paste', (e) => {
+            last_paste_content = e.clipboardData;
+        });
         editor.on('PastePreProcess', function(event) {
+            const base64_img_contents = [];
+            if (last_paste_content !== null) {
```
Comment (cedric-anne) [inline js/fileupload.js:366]:
```suggestion
            if (last_paste_content !== null && last_paste_content.types.includes('text/rtf')) {
```

**Comment by cedric-anne** [comment]
> There is still an error printed in the browser console as it tries to initially access the local file.

I think it is caused by the creation of a `DocumentFragment` based on `event.content`. Maybe it is possible to add `loading="lazy"` to `` tag in `event.content` before converting it to a fragment.

File: js/fileupload.js [hunk 3]
```diff
@@ -353,13 +353,52 @@ var insertImageInTinyMCE = function(editor, image) {
...
-                const src = image.attr('src');
+                let src = image.attr('src');
+                const file_pattern = '^file://';
+                const local_file_blob_pattern = `^blob:${CFG_GLPI.url_base}/[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}`;
+                // if file pattern or local file blob pattern
+                if (src.match(file_pattern) !== null || src.match(local_file_blob_pattern) !== null) {
+                    // Cannot modify the src attribute in this event it seems. Store the src in a data attribute.
+                    const rtf_content = base64_img_contents.shift();
+                    image.attr('data-rtf-src', `data:${rtf_content['type']};base64,` + rtf_content['content']);
+                    src = image.attr('data-rtf-src');
```
[the remark this answers is not in the document]
Comment (cconard96) [reply js/fileupload.js]: I don't know what I did differently the first time, but it is working properly now.

**Comment by cconard96** [comment]
> > There is still an error printed in the browser console as it tries to initially access the local file.
> 
> I think it is caused by the creation of a `DocumentFragment` based on `event.content`. Maybe it is possible to add `loading="lazy"` to `` tag in `event.content` before converting it to a fragment.

This isn't an issue anymore since it was changed to update the `src` attribute directly.

File: js/fileupload.js (hunk 1 above)
Comment (cedric-anne) [reply js/fileupload.js:394]: While working on #13016, I dig a little bit on this. This is really an edge case.

As far as I understand, blob URLs are produced by TinyMCE. Technically, they are valid as long as they are used in the same browser session. So the only way to have this kind of issues is to copy an image pointing to a blob from a browser A and try to use it in a browser B.

I was able to reproduce the issue by:
 - starting to write a ticket description containing a pasted image on Chrome,
 - copying the content before it was submitted,
 - pasting it a new ticket form in Firefox.

Complexing the code to handle an unlikely case is something I do not favor.

File: js/fileupload.js [hunk 4]
```diff
@@ -353,13 +353,50 @@ var insertImageInTinyMCE = function(editor, image) {
...
             // Trigger upload process for each pasted image
             var fragment = $('<div></div>');
             fragment.append(event.content);
             fragment.find('img').each(function() {
                 const image = $(this);
-                const src = image.attr('src');
+                let src = image.attr('src');
+                const file_pattern = '^file://';
+
+                if (src.match(file_pattern) !== null) {
```
Comment (cedric-anne) [inline js/fileupload.js:395]: To handle any tool that would not provide corresponding RTF content.
```suggestion
                if (src.match(file_pattern) !== null && base64_img_contents.length > 0) {
```

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

File: js/fileupload.js [hunk 5]
```diff
@@ -406,6 +443,15 @@ if (typeof tinyMCE != 'undefined') {
...
         });
+        editor.on('NodeChange', function(e) {
+            // Handle RTF image sources and remove the temporary "data-rtf-src" attribute.
+            if (e.element.tagName === 'IMG') {
+                if (e.element.hasAttribute('data-rtf-src')) {
+                    e.element.setAttribute('src', e.element.getAttribute('data-rtf-src'));
+                    e.element.removeAttribute('data-rtf-src');
+                }
+            }
+        });
```
Comment (cedric-anne) [inline js/fileupload.js:454]:
```suggestion
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**